File: typescript/Behavioral/16-Mediator/Colleague/Homeowner.ts
import { Colleague } from "./Colleague";
import { IDispatchMediator } from "../Mediator/IDispatchMediator";

export class Homeowner extends Colleague {
  constructor(mediator: IDispatchMediator, name: string) {
    super(mediator, name);
  }

  submitJobRequest(jobTitle: string, category: string): void {
    console.log(
      `\n[${this.name}] Submitting job request: '${jobTitle}' [${category}]`,
    );

    this.mediator.notify(this, "JobRequested", {
      title: jobTitle,
      category: category,
      client: this.name,
    });
  }

  receiveUpdate(message: string): void {
    console.log(`  [${this.name}] Received update: ${message}`);
  }
}

File: csharp/Behavioral/16-Mediator/Colleague/Homeowner.cs
public class Homeowner : Colleague
{
    public Homeowner(IDispatchMediator mediator, string name)
        : base(mediator, name) { }

    public void SubmitJobRequest(string jobTitle, string category)
    {
        Console.WriteLine($"\n[{Name}] Submitting job request: '{jobTitle}' [{category}]");

        Mediator.Notify(this, "JobRequested", new
        {
            Title    = jobTitle,
            Category = category,
            Client   = Name
        });
    }

    public void ReceiveUpdate(string message)
    {
        Console.WriteLine($"  [{Name}] Received update: {message}");
    }
}
